smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/validators/AwsSpecialCaseEndpointValidator.java

```diff
@@ -0,0 +1,117 @@
+import software.amazon.smithy.rulesengine.aws.traits.StandardRegionalEndpointsTrait;
+import software.amazon.smithy.utils.SetUtils;
+
+/**
+ * Validate special case endpoints from endpoint traits that are applied on a service.
+ */
+public final class AwsSpecialCaseEndpointValidator extends AbstractValidator {
+
+    private static final Set<String> SUPPORTED_PATTERNS = SetUtils.of(
+            "{region}", "{service}", "{dnsSuffix}", "{dualStackDnsSuffix}"
+    );
+
+    private static final Pattern PATTERN = Pattern.compile("\\{[^\\}]*\\}");
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardRegionalEndpointsTrait.class)) {
+            events.addAll(validateRegionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardRegionalEndpointsTrait.class)));
+        }
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(StandardPartitionalEndpointsTrait.class)) {
+            events.addAll(validatePartitionalEndpointPatterns(serviceShape,
+                    serviceShape.expectTrait(StandardPartitionalEndpointsTrait.class)));
+        }
+        return events;
+    }
+
+    private List<ValidationEvent> validateRegionalEndpointPatterns(
+            ServiceShape serviceShape, StandardRegionalEndpointsTrait regionalEndpoints) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (List<RegionSpecialCase> specialCases : regionalEndpoints.getRegionSpecialCases().values()) {
+            for (RegionSpecialCase specialCase : specialCases) {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, regionalEndpoints, specialCase.getEndpoint()));
+            }
+        }
+
+        for (List<PartitionSpecialCase> specialCases : regionalEndpoints.getPartitionSpecialCases().values()) {
+            for (PartitionSpecialCase specialCase : specialCases) {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, regionalEndpoints, specialCase.getEndpoint()));
+            }
+        }
+
+        return events;
+    }
+
+    private List<ValidationEvent> validatePartitionalEndpointPatterns(
+            ServiceShape serviceShape, StandardPartitionalEndpointsTrait partitionalEndpoints) {
+
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (List<PartitionEndpointSpecialCase> specialCases
+                : partitionalEndpoints.getPartitionEndpointSpecialCases().values()) {
+            for (PartitionEndpointSpecialCase specialCase : specialCases) {
+                events.addAll(validateEndpointPatterns(
+                        serviceShape, partitionalEndpoints, specialCase.getEndpoint()));
+            }
+        }
+
+        return events;
+    }
+
+    private List<ValidationEvent> validateEndpointPatterns(
```

Comment:
Do we need to validate that these will resolve to valid URLs given suitable substitutions for pattern values? Right now, `https://{service}..{dnsSuffix}` is valid but would fail at routing because of the `..`

Reply:
I was applying the same validation we use in validating Endpoint rules (the [RulesetUriValidator](https://github.com/smithy-lang/smithy/blob/main/smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetUriValidator.java) - which checks only the scheme).  I had originally planned to do a basic substitution (eg {region}->region or something) and validate the entire endpoint parses as a valid URI, but wanted to apply only the same validation we do for the rules.  Thoughts?